test/percy-agent-client/dom.test.ts

```diff
@@ -0,0 +1,229 @@
+// @ts-ignore
+import { check, type } from 'interactor.js'
+import * as sinon from 'sinon'
+import DOM from '../../src/percy-agent-client/dom'
+
+// Create valid DOM to pass to the DOM class.
+function createExample(dom: any) {
+  let testContainer = document.querySelector('.test-container')
+
+  if (testContainer) {
+    testContainer.remove()
+  }
+
+  testContainer = document.createElement('div')
+
+  const finalDOM = `
+    <div class="container">
+      <h1>Hello DOM testing</h1>
+      ${dom}
+    </div>
+  `
+
+  testContainer.classList.add('test-container')
+  testContainer.innerHTML = finalDOM
+  document.body.appendChild(testContainer)
+
+  return document
+}
+
+// create a stylesheet in the DOM and add rules using the CSSOM
+function createCSSOM() {
+  const style = document.createElement('style')
+  const testingContainer = document.querySelector('.test-container')
+
+  style.type = 'text/css'
+  testingContainer.appendChild(style)
+
+  const cssomStyleSheet = document.styleSheets[0] as any
+
+  cssomStyleSheet.insertRule('.box { height: 500px; width: 500px; background-color: green; }')
+}
+
+describe('DOM -', () => {
+  let dom: any
+
+  afterEach(() => {
+    dom = null
+  })
+
+  // Ensure the snapshotString method always returns a string with a doctype.
+  // The ideal way to test this is by passing a document without a doctype.
+  // We would need a custom Karma file without a doctype to test that,
+  // since the spec does not allow you to edit an existing documents doctype:
+  // "DOM level 2 doesn't support editing the document type declaration."
+  describe('snapshotString', () => {
+    beforeEach(() => {
+      dom = new DOM(document)
+    })
+
+    it('always has a doctype', () => {
+      expect(dom.snapshotString()).to.contain('<!DOCTYPE html>')
+    })
+  })
+
+  describe('passing a DOM transform option', () => {
+    let consoleStub: any
+
+    beforeEach(() => {
+      consoleStub = sinon.stub(console, 'error')
+      dom = new DOM(createExample('<span class="delete-me">Delete me</span>'), {
+        domTransformation(dom: any) {
+          const span = dom.querySelector('.delete-me')
+          span.remove()
+
+          return dom
+        },
+      })
+    })
+
+    afterEach(() => {
+      consoleStub.restore()
+    })
+
+    it('transforms the DOM', () => {
+      expect(dom.snapshotString()).to.not.contain('Delete me', 'delete-me')
+    })
+
+    it('does not modify the original DOM', () => {
+      // @ts-ignore
+      expect(document.querySelector('.delete-me').innerText).to.equal('Delete me')
+    })
+
+    // it's possible the user provides code that errors when we execute it
+    it('gracefully catches errors', () => {
+      expect(dom.snapshotString()).to.not.contain('Delete me', 'delete-me')
+      // invoke the transform function again to try and remove a non-existent element
+      expect(dom.snapshotString()).to.contain('Hello DOM testing')
+      expect(consoleStub.calledOnce).to.equal(true)
```

Comment:
Add an expect before the first `snapshotString` method call to ensure `console.error` has not been called.